sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_async.py

```diff
@@ -203,3 +194,14 @@ def _message_received(self, message: uamqp.Message) -> None:
                         last_exception,
                     )
                     raise last_exception
+
+        while not self._event_queue.empty():
+            message = self._event_queue.get()
+            event_data = EventData._from_message(  # pylint:disable=protected-access
+                message
+            )
+            self._last_received_event = event_data
+            trace_link_message(event_data)
+            await self._on_event_received(event_data)
+            self._event_queue.task_done()
+        return
```

Comment:
Moving the message processing out here is probably a good idea - in fact we could probably move it to a separate function - however we now no longer have any error processing in the scenario that loading a message doesn't work.
I'm thinking specifically of:
```
if self._last_received_event:
    self._offset = self._last_received_event.offset
```
Do you think this is needed?

Reply:
emmm, What is `loading a message doesn't work`?

`self._offset` is used for specifying starting position.
- When exception happens, uamqp receiver will start from where it fails - last received event.
- In the case of uamqp receiver fails at the very beginning (without receiving any messages), the offset remains unchanged.
- I use `_last_received_event` instead of setting `_offset` directly because of the performance concern -- lazy initialization.

Reply:
For "loading a message doesn't work" I mean that the event data received is somehow invalid and the line `EventData._from_message(event)` raises an error.

So we don't need to track `self._offset` in normal circumstances?

Reply:
aha, got what you mean.

If `EventData._from_message(event)` raises an error, it will bubble up to the eventprocessor, eventprocessor then will call `on_error`, close and re-open the internal partition receiver.

I think it's the expected behavior.

----
we need to track the self._offset, we track `self._last_received_event` which contains the last offset information. (But I don't think this is your major concern)